pkgs/by-name/je/jextract/package.nix

```diff
@@ -46,8 +46,10 @@ stdenv.mkDerivation {
     description = "Tool which mechanically generates Java bindings from a native library headers";
     mainProgram = "jextract";
     homepage = "https://github.com/openjdk/jextract";
-    platforms = jdk22.meta.platforms;
+    platforms = jdk23.meta.platforms;
     license = licenses.gpl2Only;
     maintainers = with maintainers; [ jlesquembre sharzy ];
+    # Not yet updated for JDK 23
+    broken = true;
```

Comment:
It would be more respectful to jextract users to wait with the jdk22 removal

Reply:
I think keeping older versions of Java that are possible insecure is more of an issue than having one specific package to break. We could just have marked the JDK22 (and all packages that depend on it) as insecure instead and keep it for longer, but I concur that we should, if anything, to reduce the number of Java packages in the tree to make sure that the ones that we have are up to date.

Reply:
The `jextract` for the LTS JDK 21 is still in‐tree. There’s been seemingly no movement in the `jextract` repository porting it to an on‐EOL JDK, so I don’t know what’s going on there. The same upstream that maintains `jextract` sets the OpenJDK support policy and since it is guaranteed that we will get these versions going EOL every release (when it’s not time for a new LTS) we can’t let the list of insecure, unsupported JDKs keep growing indefinitely, as happened before the 24.11 cycle. I would prefer we move to a scheme like Fedora where the latest JDK is available as `openjdk_latest` and simply rolls over to the new version when released.

Reply:
> older versions of Java that are possibly insecure

Well all software is possibly insecure. I know that you mean unmaintained, which is similar.

> We could just have marked the JDK22 (and all packages that depend on it) as insecure instead and keep it for longer

That would seem like a better solution to me.

> reduce the number of Java packages in the tree to make sure that the ones that we have are up to date

I agree with this as well. But I would suggest something like what I see for [java in Debian Trixie](https://packages.debian.org/trixie/java/) where they support:

* 17
* 21
* 22
* 23
* 24-ea

At an absolute minimum provide: LTS versions, the latest release, the previous release. It would also be nice to see EA versions installable through Nix, but obviously someone has to do the work.

Releases that are unmaintained upstream could be marked as unsupported/insecure/deprecated, just so long as projects aren't broken without a few months warning.

In the case of `jextract`, which is a tool to build bindings for Java's Foreign Function & Memory API which went final in JDK 22. The only way (I'm pretty sure) to build bindings for the final API is with the JDK 22-based jextract. So, at this point, there is no way to do that using nixpkgs.

Reply:
> The `jextract` for the LTS JDK 21 is still in‐tree.

As I mentioned in my reply above, this version is not very useful because it is for a non-final FFM API and requires use of JDK 21 with "preview features" enabled.

> There’s been seemingly no movement in the `jextract` repository porting it to a non‐EOL JDK

This appears to be the case and is disappointing. Maybe someone should reach out to the OpenJDK project and ask about it.

> we can’t let the list of insecure, unsupported JDKs keep growing indefinitely

I agree. But hopefully there's a way to do this with a little more warning (e.g. a 3-6 months) to dependent projects.

Reply:
Honestly if we wanted to diverge from the Fedora policy of LTSes + rolling latest release then I’d prefer simply skipping the 6‐month releases entirely. They go EOL like clockwork so software choosing to live on the bleeding edge should be prepared to roll over to the next release when it comes out – that’s the price of relying on features in releases with a very limited support lifespan. Why a first‐party project like `jextract` can’t keep up with that, I don’t know; you’d presumably have to ask the OpenJDK developers. It’s already a huge pain for Nixpkgs to deal with software with such a limited lifespan that isn’t perfectly synchronized to our release cycle.

Note that we usually carry only one version of software in Nixpkgs; OpenJDK is already very high up the list of packages we carry a large number of versions for, with e.g. more versions than Node.js and the same number as CPython 3. We mark EOL language implementations as insecure and consider them high priorities for removal because they are highly complex pieces of software with a huge attack surface that has a downstream security impact on every package that uses them. Just look at the [OpenJDK Vulnerability Advisories](https://openjdk.org/groups/vulnerability/advisories/) to see how regularly security issues are found and fixed. EOL JDKs don’t get those fixes.

Reply:
> Well all software is possibly insecure. I know that you mean unmaintained, which is similar.

Well, looking at the latest advisory for Java it seems there are new security issues affecting all supported Java versions. It is very unlikely this doesn't also affect JDK 22 so we actually know it has security issues: https://openjdk.org/groups/vulnerability/advisories/2024-10-15.

Those security advisories are released every 3 months and almost always have at least one security issue, so for a version that is maintained by 9 months like our stable branches we are unlikely to ship the previous version and keep it secure.

I actually like the proposal from @emilazy here to keep all LTS versions+`openjdk_latest` with the last version, because at least we can ensure that it is possible to keep our JDK versions up to date in stable.

Reply:
FWIW I spent about 30 minutes looking into a way to patch `jextract` for the latest JDK and I didn’t succeed, but I suspect someone with more familiarity would have an easier time with it. It didn’t seem like it should be impossible by any means; it just exceeded the time‐box I had allocated to it. If we could do that then it could also be sent upstream to OpenJDK.

> I agree. But hopefully there's a way to do this with a little more warning (e.g. a 3-6 months) to dependent projects.

The OpenJDK release and support cycle is fixed – the 6 months warning is when a new non‐LTS release comes out, and the 3 months warning is 3 months after that. That cycle is on the OpenJDK project; we don’t have the resources to support them longer than upstream does. We’ve regularly lagged behind on security‐critical updates even to the versions we currently have, and 24.05 was in a much worse state (I think all the JDKs there should actually be marked as `knownVulnerabilities`).

I believe our stable branches are actually maintained for 7 months? But: that means that OpenJDK 23 is going to go EOL during the 24.11 cycle, which is really awkward. Either we mark it with `knownVulnerabilities` and have only programs using LTS JDKs working without a security version for the last few months of the release cycle, or we backport a bump to the next JDK and potentially break other software in the process. I believe long‐term support distributions like RHEL don’t even bother shipping non‐LTS JDKs. It’s really something you should only depend on if you can be sufficiently agile at supporting upcoming versions, given the upstream policy.

Reply:
[This mailing list message](https://mail.openjdk.org/pipermail/jextract-dev/2024-September/001948.html) claims that `jextract` should basically “just work” with JDK 23. That was unfortunately not my experience, but it could just be build system issues I’d be running into, so if someone wanted to have a go at getting the package unbroken I suspect it shouldn’t be too difficult.

Reply:
The arguments about security are compelling. And I know this is a volunteer project with limited resources. I am new to Nixpkgs and am trying to contribute as much as I can (e.g. https://github.com/NixOS/nixpkgs/pull/271127), but am not (yet) ready to try to (help) support OpenJDK itself.

I have been trying to replace (as much as possible) Homebrew and [SDKMAN!](https://sdkman.io) with Nixpkgs.

I personally prefer to focus on getting `jextract` updated: https://github.com/NixOS/nixpkgs/pull/271127

But I also want to share my experiences a a Nixpkgs user and having something like `jdk22` disappear so quickly is not very friendly to downstream projects. A difference between installing JDK binaries with Homebrew or SDKMAN! and using Nix flakes is that dropping JDK 22 will actually break Nix builds, whereas Homebrew or SDKMAN! will allow you to leave your old (unmaintained/insecure) JDK installed and usable.

Reply:
FWIW, it’s possible to pin the last Nixpkgs version with OpenJDK 22, and use that package alongside other packages on your system, thanks to Nix’s hermetic pinning of dependency trees. That’s of course inconvenient, and in some ways slightly worse than if we were carrying an insecure‐marked OpenJDK 22 in the current tree (namely: its dependencies won’t get updates either, though since the JDK itself is probably the biggest source of vulnerabilities in its dependency closure that might not matter too much), but hopefully it can at least serve at a pinch when you really can’t do without the old version. (It also avoids having to build the JDK yourself, since we don’t spend Hydra build farm resources on insecure packages.)

I realize it’s frustrating from a user point of view, though, and I sympathize with the pain of dealing with short support cycles – ideally OpenJDK would provide an overlapping support window for their non‐LTS releases, but they don’t. I think they would have the ability to do that; we support a one month security support overlap between our own stable versions despite having considerably fewer resources than Oracle. I don’t really know why they don’t; I was wondering if perhaps it’s because they sell extended commercial support, but that’s only for the LTS releases. I get the impression that they consider their non‐LTS releases to essentially be previews for the next LTS.

If you want to try updating `jextract` to the latest early access build and seeing if it’ll work on JDK 23 that would be great, and please do ping me for review! I can try to help if you run into any issues as well. I didn’t want to mark it as broken here but making sure we had the latest supported JDK and weren’t at risk of shipping an EOL one in 24.11 took priority. If we could get it fixed before the 24.11 release I’d be very happy.